Fixes for #2803 Cancel nearby upload -> Next upload gets the same title/description

SamuelleMollard opened the pull request:
**Description (required)**
My team and I fixed this bug : #2803.

Fixes #2803 Cancel nearby upload -> Next upload gets the same title/description

What changes did you make and why?

I added a function to clear the NearbyParentFragment cache and modified MainActivity to call that function in onDestroy, through an other function I defined in MainActivity which does the call on the right instance

**Tests performed (required)**

Tested on prod on Asus Zenpad with API level 24.

maskaravivek commented:
@SamuelleMollard Can you explain how this PR fixes #102.

SamuelleMollard commented:
Um, looking at it I think they forgot to push their modifications so it doesn't.
I'll talk about it with them, but as of now, since the only modified files are the ones I changed it doesn't

maskaravivek commented:
In fact, it is good that this PR doesn't fix both the issues. Ideally, a PR should fix just 1 issue. :)

Feel free to create a separate PR for other issues. :)

SamuelleMollard commented:
Very well, I will. 
I'm changing the description then

SamuelleMollard commented:
It is updated, tell me if there is anything wring with it so that I can fix it
I will see with my team so that we do this other pr at a later time

maskaravivek commented on app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java:
```diff
@@ -435,12 +435,32 @@ protected void onResume() {
+    /**
+     * This function will clear the cache made by the Nearby feature
+     */
+    public void clearNearbyCache() {
```
Use the MVP pattern.

SamuelleMollard replied:
Ok, I moved the code in NearbyParentFragment in NearbyController and I made the NearbyParentFragment's method only call the controller's.

I didn't change the MainActivity one since it just finds the right fragment to be able to call the function

neslihanturan replied:
Instead of finding the fragment, you should access the singletone presenter object and implement your method there, ideally.

neslihanturan replied:
There are some examples for how to do in MainActivity, you can check them.

SamuelleMollard commented:
This should be fine now, please tell me if there is any other problem.

neslihanturan commented on app/src/main/java/fr/free/nrw/commons/nearby/NearbyController.java:
```diff
@@ -245,6 +246,14 @@ public NearbyPlacesInfo loadAttractionsFromLocation(LatLng curLatLng, LatLng sea
+     * A function which clears the shared preferences
+     * Used to fix #3202
+     */
+    public void removeSharedPrefs(JsonKvStore applicationKvStore, String PLACE_OBJECT) {
```
This method seemed to me like an utility method and NearbyController is mainly to control fetching places from wikidata servers. Would be nice if you can move this to some approptiate utility class.

neslihanturan commented on app/src/main/java/fr/free/nrw/commons/nearby/fragments/NearbyParentFragment.java:
```diff
@@ -974,6 +974,14 @@ private void storeSharedPrefs(Place selectedPlace) {
+     * This function clears the place data from the application, so that when the user leaves the app and
+     * tries to upload an other image the data is not kept (see #2803)
+     */
+    public void removeSharedPrefs() {
```
This also can be implemented in NearbyParentPresenter, instead of fragment. Since this is not a UI related task and according to MVP it should not be implemented here.

SamuelleMollard commented:
I currently can't call the controller because there is an object it requires and that object is only owned by the fragment (hte JSonKvStore object). Fixing that may require a few hours of refactoring and i don't have the time right now.
I will try and do it Thursday afternoon, but if I don't have time to do it by then, I probably won't so it at all since I have a lot of other things to do and just don't have the time. I will push what I will have done if it works though.

neslihanturan commented:
You can always pass this object as parameter and I think you can access the object from the activity too. so no need to extra refactor anything.

SamuelleMollard commented:
Ok, in any case I will do that on Thursday, because I have way too much work before that. If it's this straightforward, there is no reason i would not be able to do it.

SamuelleMollard commented:
I might try and do that today though if it's this straightforward though, since I may have a little bit of time
Depends on other people too though so I'm not sure

SamuelleMollard commented:
I have tried, but the JSonKvStore object is not reachable neither in MainActivity not in NearbyParentFragmentPresenter, and I have tried to inject it to the latter, but the app just end up crashing. I'll probably have a look on Thursday.

SamuelleMollard commented:
Ok, I added a getter for the applicationKvStore object, added that in the interface so that the presenter can access it, and implemented the method there. I ran the same tests and the bug is fixed and the app doesn't crash, unlike my attempts at using injection. I pushed the changes, tell me if it's good or not for you. If it is not, I will probably give up and write a document for the next person working on this issue to start from what I have done, so that they don't have to start from scratch

SamuelleMollard commented:
Very well, thanks

ashishkumar468 commented on app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java:
```diff
@@ -435,12 +435,15 @@ protected void onResume() {
         locationManager.unregisterLocationManager();
         // Remove ourself from hashmap to prevent memory leaks
         locationManager = null;
+        NearbyParentFragmentPresenter.getInstance().removeNearbyPreferences();
```
Do this in onStop(), onDestroy is not to be relied upon

SamuelleMollard replied:
If I do this on onStop(), it will cause unwanted behaviour (not filling the fields when the user leaves the app after getting in the photo activity and comes back and takes the photo). Are you sure you want me to move that line ?

ashishkumar468 replied:
AFAIF, edit text retain their states in such cases, if because of anything which we have done, they do not, we can save use savedInstanceState to retain such things

SamuelleMollard replied:
To be honest, my tablet's broken right now so I can't really test anything new, plus I don't think I will be working more on this since I have other projects coming up. I also don't really see how you would do what you are describing. If you want me to just add a 'onStop()' function and move my line there I can do it, but that's about all I can really do, and as I said earlier, I doubt it would be a good solution.
Finally, I don't quite understand why we should not be relying on onDestroy(), especially since the code already makes use of it.

ashishkumar468 commented on app/src/main/java/fr/free/nrw/commons/nearby/NearbyController.java:
```diff
@@ -20,6 +20,7 @@
 import javax.inject.Inject;
 
 import fr.free.nrw.commons.R;
+import fr.free.nrw.commons.kvstore.JsonKvStore;
```
Please revert the formatting changes

SamuelleMollard replied:
Done

neslihanturan commented on app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java:
```diff
@@ -438,12 +438,15 @@ protected void onResume() {
         setNotificationCount();
     }
 
+
```
Please revert extra spaces

neslihanturan commented on app/src/main/java/fr/free/nrw/commons/nearby/NearbyController.java:
```diff
@@ -246,6 +246,7 @@ public NearbyPlacesInfo loadAttractionsFromLocation(LatLng curLatLng, LatLng sea
         return baseMarkerOptions;
     }
 
+
```
also here:)

neslihanturan commented on app/src/main/java/fr/free/nrw/commons/nearby/presenter/NearbyParentFragmentPresenter.java:
```diff
@@ -376,4 +377,10 @@ public boolean areLocationsClose(LatLng cameraTarget, LatLng lastKnownLocation)
         }
     }
+
+
```
also here, one

SamuelleMollard commented:
I'm really sorry, but, like, my tablet is literally dead, and I'm not buying a new one before long, there is no way for me to test anything at this point.
I CAN clone it, and remove spaces with gedit, but at this point I kind of feel like there's really no point.
Please either accept the PR or copy the changes, fix what you want to fix and merge as your own PR, as I don't think I'm going to be working much more on this project.
I do not really care about being credited or anything, don't worry about it, if you want to merge it as your own PR that's fine by me

neslihanturan commented:
Oh sorry to hear that @SamuelleMollard , you can always use an emulator if you like though. Thank you for the contribution, I am sure someone will reclaim this PR.

SamuelleMollard commented:
I can't use an emulator, my computer already barely has enough RAM to use Android Studio
No problem, I hope your project goes well

neslihanturan commented:
Since the PR owner said they are not interested anymore, closing this one. Feel free to get help from here to find a solution the same bug :)
